github/basic_auth_token.go

```diff
@@ -0,0 +1,199 @@
+//
+type Creds struct {
+	Username       string
+	Password       string
+	PostBodyReader io.Reader
+	Config         *oauth2.Config
+}
+
+// Token set username/password and postbody and do basic auth
+func (ba Creds) Token() (tk *oauth2.Token, err error) {
+
+	return TokenSourceBasicAuth(ba)
+
+}
+
+// BasicAuthRequestBody is the struct for generating the body for the authentication POST
+//
+// its used to generate the postBodyReader that goes into the Creds with the username/password
+type BasicAuthRequestBody struct {
+	ClientId     string   `json:"client_id"`
+	ClientSecret string   `json:"client_secret"`
+	Note         string   `json:"note"`
+	Scopes       []string `json:"scopes"`
+}
+
+// tokenBasicAuthJSON structure for deserializing the basic auth json response
+type tokenBasicAuthJSON struct {
+	Token       string `json:"token"`
+	Url         string `json:"url"`
+	AppRegistry struct {
+		Url      string `json:"url"`
+		ClientID string `json:"client_id"`
+	} `json:"app_registry"`
+	Note         string `json:"note"`
+	RefreshToken string `json:"refresh_token"`
+}
+
+// NewBasicAuthClient returns an http client that will do the Basic Auth call to github to get a token
+// via the user creds specified
+//
+// The returned http.Client can be passed to github.NewClient
+//
+// Sample Usage:
+//
+//	oa2 := oauth2.Config{ ... }
+//
+//	tc, err := github.NewBasicAuthClient(oa2, "zup", "nuch", "my oauth", []string{"public_repo"})
+//
+//	client := github.NewClient(tc)
+//
+// the github api docs are here:
+//
+//	https://developer.github.com/v3/oauth_authorizations/#create-a-new-authorization
+//
+func NewBasicAuthClient(oa2 oauth2.Config, username, password, note string, repos []string) (tc *http.Client, err error) {
+
+	postBody := BasicAuthRequestBody{
+		oa2.ClientID,
+		oa2.ClientSecret,
+		note,
+		repos,
+	}
+
+	pb, err := json.Marshal(postBody)
+	if err != nil {
+		return
+	}
+
+	creds := Creds{username, password, bytes.NewReader(pb), &oa2}
+
+	tc = oauth2.NewClient(context.Background(), creds)
+
+	return
+
+}
+
+// TokenSourceBasicAuth - use Basic Auth (username/password) to get a token
+func TokenSourceBasicAuth(creds Creds) (*oauth2.Token, error) {
+
+	c := creds.Config
+
+	tk, err := retrieveTokenBasicAuth(c.Endpoint.TokenURL, creds.Username, creds.Password, creds.PostBodyReader)
+	if err != nil {
+		return nil, err
+	}
+
+	if tk == nil {
+		return &oauth2.Token{}, fmt.Errorf("got an empty token")
+	}
+
+	return &oauth2.Token{
+		AccessToken:  tk.AccessToken,
+		TokenType:    tk.TokenType,
+		RefreshToken: tk.RefreshToken,
+		Expiry:       tk.Expiry,
+	}, nil
+
+}
+
+// retrieveTokenBasicAuth - Do Basic auth via username/password not client_id/client_secret
+//
+// POST body for the auth call comes from the caller as a Reader
+func retrieveTokenBasicAuth(TokenURL, Username, Password string, postBodyReader io.Reader) (*oauth2.Token, error) {
+
+	hc := http.Client{}
```

Comment:
this is a problem.  All API calls have to be made with the client that is provided to `NewClient`.  That's because a zero-value `http.Client` won't work in certain environments, notable App Engine as well as Google's prod environment.